experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,2237 @@
+    for (unsigned i = 0; i < size; ++i)
+      ins.push_back(getState<ConsumerRW>(op->getOperand(i), subset));
+
+    for (unsigned i = 0; i < size; ++i)
+      insData.emplace_back(ins[i]);
+    mergeTEHB.datawidth = outsData.dataWidth;
+  }
+
+  void reset() override {
+    if (outsData.hasValue())
+      resetDataFull(ins, outs, insData, outsData);
+    else
+      resetDataless(ins, outs);
+  }
+
+  void exec(bool isClkRisingEdge) override {
+    if (outsData.hasValue())
+      execDataFull(isClkRisingEdge, ins, outs, insData, outsData);
+    else
+      execDataless(isClkRisingEdge, ins, outs);
+  }
+
+  void printStates() override {
+    for (unsigned i = 0; i < size; ++i)
+      printValue<ConsumerRW, const Data>("ins", ins[i], insData[i].data);
+    printValue<ProducerRW, Data>("outs", outs, outsData.data);
+  }
+
+private:
+  handshake::MergeOp op;
+
+  // parameters
+  unsigned size;
+
+  // ports
+  std::vector<ConsumerRW *> ins;
+  ProducerRW *outs;
+
+  std::vector<ConsumerData> insData;
+  ProducerData outsData;
+
+  // merge dataless
+  bool tehbValid = false, tehbReady = false;
+  ConsumerRW insTEHB;
+  // merge datafull
+  Data tehbDataIn;
+
+  // internal components
+  TEHBSupport mergeTEHB;
+
+  void resetDataless(std::vector<ConsumerRW *> &ins, ProducerRW *outs) {
+    tehbValid = false;
+    for (unsigned i = 0; i < size; ++i)
+      tehbValid = tehbValid || ins[i]->valid;
+
+    for (unsigned i = 0; i < size; ++i)
+      ins[i]->ready = tehbReady;
+
+    mergeTEHB.resetDataless(&insTEHB, outs);
+  }
+
+  void execDataless(bool isClkRisingEdge, std::vector<ConsumerRW *> &ins,
+                    ProducerRW *outs) {
+    tehbValid = false;
+    for (unsigned i = 0; i < size; ++i)
+      tehbValid = tehbValid || ins[i]->valid;
+
+    for (unsigned i = 0; i < size; ++i)
+      ins[i]->ready = tehbReady;
+
+    mergeTEHB.execDataless(isClkRisingEdge, &insTEHB, outs);
+  }
+
+  void resetDataFull(std::vector<ConsumerRW *> &ins, ProducerRW *outs,
+                     std::vector<ConsumerData> &insData,
+                     ProducerData &outsData) {
+    tehbValid = false;
+    tehbDataIn = *insData[0].data;
+    for (unsigned i = 0; i < size; ++i) {
+      if (ins[i]->valid) {
+        tehbValid = true;
+        tehbDataIn = *insData[i].data;
+        break;
+      }
+    }
+
+    for (unsigned i = 0; i < size; ++i)
+      ins[i]->ready = tehbReady;
+
+    mergeTEHB.reset(&insTEHB, outs, &tehbDataIn, outsData.data);
+  }
+
+  void execDataFull(bool isClkRisingEdge, std::vector<ConsumerRW *> &ins,
+                    ProducerRW *outs, std::vector<ConsumerData> &insData,
+                    ProducerData &outsData) {
+    tehbValid = false;
+    tehbDataIn = *insData[0].data;
+    for (unsigned i = 0; i < size; ++i) {
+      if (ins[i]->valid) {
+        tehbValid = true;
+        tehbDataIn = *insData[i].data;
+        break;
+      }
+    }
+    for (unsigned i = 0; i < size; ++i)
+      ins[i]->ready = tehbReady;
```

Comment:
It seems that you can just call `resetDataFull` to do all of this no? That removes some code duplication.